src/theory/arith/theory_arith.cpp

```diff
@@ -123,7 +132,50 @@ void TheoryArith::propagate(Effort e) {
 }
 bool TheoryArith::collectModelInfo(TheoryModel* m)
 {
-  return d_internal->collectModelInfo(m);
+  std::set<Node> termSet;
+  // Work out which variables are needed
+  const std::set<Kind>& irrKinds = m->getIrrelevantKinds();
+  computeAssertedTerms(termSet, irrKinds);
+  // this overrides behavior to not assert equality engine
+  return collectModelValues(m, termSet);
+}
+
+bool TheoryArith::collectModelValues(TheoryModel* m,
+                                     const std::set<Node>& termSet)
+{
+  // get the model from the linear solver
+  std::map<Node, Node> arithModel;
+  d_internal->collectModelValues(termSet, arithModel);
+  // if non-linear is enabled, intercept the model, which may repair its values
+  if (d_nonlinearExtension != nullptr)
+  {
+    // Non-linear may repair values to satisfy non-linear constraints (see
+    // documentation for NonlinearExtension::interceptModel).
+    d_nonlinearExtension->interceptModel(arithModel);
+  }
+  // We are now ready to assert the model.
+  for (std::pair<const Node, Node>& p : arithModel)
+  {
+    if (m->assertEquality(p.first, p.second, true))
+    {
+      continue;
+    }
+    // If we failed to assert an equality, it is likely due to theory
+    // combination, namely the repaired model for non-linear changed
+    // an equality status that was agreed upon by both (linear) arithmetic
+    // and another theory. In this case, we must add a lemma, or otherwise
+    // we would terminate with an invalid model. Thus, we add a splitting
+    // lemma of the form ( x = v V x != v ) where v is the model value
+    // assigned by the non-linear solver to x.
+    if (d_nonlinearExtension != nullptr)
+    {
+      Node eq = p.first.eqNode(p.second);
+      Node lem = NodeManager::currentNM()->mkNode(kind::OR, eq, eq.negate());
+      d_out->lemma(lem);
```

Comment:
This becomes interesting when CAD constructs a non-rational model... we should expect this to explode at some point :-)
Maybe add an assertion already that `!eq.isNull()` or something like that?

Reply:
Do you mean to check that p.first / p.second are not null? (eq is obvious not null)

Reply:
I'm not entirely sure how all this works together, but CAD models may come as witness terms and thus `p.first` may be an arithmetic variable while `p.second` may be a Boolean formula. Unless witness terms are handled differently... but I don't really see that anywhere.

Reply:
We should always ensure p.first and p.second have the same type, through witness.

This map is passed through to NlModel, updated possibly with witness terms here: https://github.com/CVC4/CVC4/blob/master/src/theory/arith/nl/nl_model.cpp#L1260.